stdlib/public/core/UnsafeUninitializedPointer.swift

```diff
@@ -0,0 +1,140 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2021 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+/// Provides scoped access to a raw buffer pointer with the specified byte count
+/// and alignment.
+///
+/// - Parameters:
+///   - byteCount: The number of bytes to temporarily allocate. `byteCount` must
+///     not be negative.
+///   - alignment: The alignment of the new, temporary region of allocated
+///     memory, in bytes.
+///   - body: A closure to invoke and to which the allocated buffer pointer
+///     should be passed.
+///
+///  - Returns: Whatever is returned by `body`.
+///
+///  - Throws: Whatever is thrown by `body`.
+///
+/// This function is useful for cheaply allocating raw storage for a brief
+/// duration. Storage may be allocated on the heap or on the stack, depending on
+/// the required size and alignment.
+///
+/// When `body` is called, the contents of the buffer pointer passed to it are
+/// in an unspecified, uninitialized state. `body` is responsible for
+/// initializing the buffer pointer before it is used _and_ for deinitializing
+/// it before returning. `body` does not need to deallocate the buffer pointer.
+///
+/// The implementation may allocate a larger buffer pointer than is strictly
+/// necessary to contain `byteCount` bytes. The behavior of a program that
+/// attempts to access any such additional storage is undefined.
+///
+/// The buffer pointer passed to `body` (as well as any pointers to elements in
+/// the buffer) must not escape—it will be deallocated when `body` returns and
+/// cannot be used afterward.
+@_transparent
+@available(macOS 9999, iOS 9999, watchOS 9999, tvOS 9999, *)
+public func withUnsafeUninitializedMutableRawBufferPointer<ReturnType>(byteCount: Int, alignment: Int, _ body: (UnsafeMutableRawBufferPointer) throws -> ReturnType) rethrows -> ReturnType {
+  _debugPrecondition(byteCount > 0, "Too little uninitialized memory requested.")
+  _debugPrecondition(alignment > 0, "Nonsensical alignment requested.")
+
+  // FIXME: Builtin.stackAlloc() should return an Optional<Builtin.RawPointer>.
+  let stackPointer = Builtin.stackAlloc(byteCount._builtinWordValue, alignment._builtinWordValue)
+  if Int(Builtin.ptrtoint_Word(stackPointer)) != 0 {
+    defer { _fixLifetime(stackPointer) }
+    return try body(.init(start: UnsafeMutableRawPointer(stackPointer), count: byteCount))
```

Comment:
There's a comment in [Runtime.swift](https://github.com/apple/swift/blob/5f41f94f42c9ddd56a18ae1e6c685c7fa5bb32b4/stdlib/public/core/Runtime.swift#L30):

> We use Builtin.Word here because Builtin.RawPointer can't be nil.

```suggestion
  let stackPtrOrNil: Builtin.Word = Builtin.stackAlloc(
    byteCount._builtinWordValue, alignment._builtinWordValue)
  if let start = UnsafeMutableRawPointer(bitPattern: Int(stackPtrOrNil)) {
    return try body(.init(start: start, count: byteCount))
```

Why is the `_fixLifetime` call needed?

Reply:
I am aware of that comment, thanks! I am waiting for feedback from the core team regarding the correct way to express a nullable pointer in a builtin. :)

> Why is the _fixLifetime call needed?

To prevent the compiler from reclaiming the stack space occupied by the temporary allocation. See other uses of this idiom in the stdlib. (It's possible it's not needed here, but again am waiting for core team feedback.)

Reply:
The returned value is of type `Builtin.RawPointer`. The last use of that value is _before_ the call to `body()` returns, because it's passed to `UnsafeMutableRawPointer.init(_:)`. Under strict lifetime rules, it is then eligible for destruction before `body()` is even called.

Reply:
@benrimmington I attempted your proposed change here but it doesn't actually compile, so I'm backing it out. I'm going to wait for feedback from the core team to see how best to implement this pattern. :)